controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,260 @@
+	// management controller.
+	var bmcCreds *bmc.Credentials
+
+	bmcCreds, _, err = r.buildAndValidateBMCCredentials(request, host)
+	if err != nil || bmcCreds == nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to get bmc secret")
+	}
+
+	prov, err := r.ProvisionerFactory.NewProvisioner(provisioner.BuildHostData(*host, *bmcCreds), nil)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	ready, err := prov.IsReady()
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to check services availability")
+	}
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	httpHeaders, err := r.getHTTPHeaders(*subscription)
+
+	if err != nil {
+		reqLogger.Error(err, "failed to get http headers")
+		return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if subscription.Status.SubscriptionID == "" {
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to add a subscription")
+			}
+
+			// Update the subscription with the new ID
+			err := r.Status().Update(ctx, subscription)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+
+		} else {
+			// Update an existing subscription by recreating it in the BMC
+
+			originalSubscription := subscription.DeepCopy()
+
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription")
+			}
+
+			if _, err := prov.RemoveBMCEventSubscriptionForNode(*originalSubscription); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to remove a subscription")
+			}
+
+			return ctrl.Result{}, nil
+		}
+	} else {
+		// being deleted
+		if subscriptionHasFinalizer(subscription) {
+			if _, err := prov.RemoveBMCEventSubscriptionForNode(*subscription); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to remove a subscription")
+			}
+
+			// Remove finalizer to allow deletion
+			subscription.Finalizers = utils.FilterStringFromList(
+				subscription.Finalizers, metal3v1alpha1.BMCEventSubscriptionFinalizer)
+			reqLogger.Info("cleanup is complete, removed finalizer",
+				"remaining", subscription.Finalizers)
+			if err := r.Update(context.Background(), subscription); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	return
+
+}
+
+func (r *BMCEventSubscriptionReconciler) getBMCSecret(request ctrl.Request, host *metal3v1alpha1.BareMetalHost) (bmcCredsSecret *corev1.Secret, err error) {
+	secretKey := host.CredentialsKey()
+	bmcCredsSecret = &corev1.Secret{}
```

Comment:
nit: why return a pointer? you dereference it eventually anyway (and `nil` should not be an option)

Reply:
Following the lead from [our old bmh controller implementation](https://github.com/metal3-io/baremetal-operator/blob/388bcf6f7de45c1889b89918c29ed1dbb06a7108/controllers/metal3.io/baremetalhost_controller.go#L1218)

Should we refactor this to use the newish secret manager?